src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/VisualStudioWorkspace_OutOfProc.cs

```diff
@@ -81,5 +81,23 @@ public void SetFeatureOption(string feature, string optionName, string language,
             => _inProc.SetFeatureOption(feature, optionName, language, valueString);
 
         public string GetWorkingFolder() => _inProc.GetWorkingFolder();
+
+        /// <summary>
+        /// Waits for a WorkspaceChange event that happens after the code <paramref name="action"/> has ran. This should
+        /// only be needed if there is asynchrony that is untracked by Roslyn waiters; for example, modifying files and waiting
+        /// for file watchers or doing some action that triggers an asynchronous project system update.
+        /// </summary>
+        /// <param name="changeKind">The type of WorkspaceChange to look for.</param>
+        /// <param name="action">The action to run.</param>
+        public void WaitForWorkspaceChangeFromAction(WorkspaceChangeKind changeKind, int numberOfChanges, Action action)
+        {
+            _inProc.BeginWaitForWorkspaceChange(changeKind, numberOfChanges);
+            action();
+            _inProc.EndWaitForWorkspaceChange(Helper.HangMitigatingTimeout);
```

Comment:
:bulb:

```
int currentChangeCount = _inProc.GetTotalChangeCount(changeKind);
action();
_inProc.WaitForTotalChangeCountToReach(Helper.HangMitigatingTimeout, changeKind, currentChangeCount + numberOfChanges);
```

Reply:
I guess I wanted to establish the pattern if you're also filtering in some fancier way. Which actually gives me a better idea: no reason to pass change kind in, just pass a `Predicate<WorkspaceChangeEventArgs>` so if down the road you want to filter in some fancier way (say, on a particular ID) you're good too.